src/celeritas/ext/detail/G4VecgeomConverter.hh

```diff
@@ -0,0 +1,147 @@
+//----------------------------------*-C++-*----------------------------------//
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+//---------------------------------------------------------------------------//
+/*!
+ * \file G4VecgeomConverter.hh
+ * \brief Class to create a VecGeom model from a pre-existing Geant4 geometry
+ *
+ * Original code from G4VecGeomNav package by John Apostolakis et.al.
+ *
+ * Original source:
+ * https://gitlab.cern.ch/VecGeom/g4vecgeomnav/-/raw/fdd310842fa71c58b3d99646159ef1993a0366b0/include/G4VecGeomConverter.h
+ */
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <VecGeom/base/TypeMap.h>
+#include <VecGeom/management/GeoManager.h>
+
+#include "Geant4/G4PVReplica.hh"
+#include "Geant4/G4RotationMatrix.hh"
+#include "Geant4/G4ThreeVector.hh"
+
+class G4LogicalVolume;
+class G4AffineTransformation;
+class G4VPhysicalVolume;
+class G4VSolid;
+
+namespace celeritas
+{
+inline namespace VECGEOM_IMPL_NAMESPACE
+{
+class LogicalVolume;
+class Transformation3D;
+class UnplacedBox;
+class VUnplacedVolume;
+class VPlacedVolume;
+
+} // namespace VECGEOM_IMPL_NAMESPACE
+} // namespace vecgeom
+
+// class converting G4 to VecGeom (only geometry; no materials)
+class G4VecGeomConverter
+{
+  private:
+    /** Remember pointer to generated world from imported G4 geometry. */
+    vecgeom::VPlacedVolume const* fWorld;
+
+    // vecgeom::BidirectionalTypeMap<unsigned int, G4VPhysicalVolume const*>
+    // fPlacedVolumeMap; using G4PVReplicaPair = std::pair<G4VPhysicalVolume
+    // const*, int>; vecgeom::BidirectionalTypeMap<unsigned int,
+    // G4PVReplicaPair> fPlacedVolumeMap;
+
+    // one G4 physical volume can correspond to multiple vecgeom placed volumes
+    // (in case of replicas)
+    vecgeom::BidirectionalTypeMap<std::vector<vecgeom::VPlacedVolume const*> const*,
+                                  G4VPhysicalVolume const*>
+        fPlacedVolumeMap;
+
+    vecgeom::BidirectionalTypeMap<vecgeom::VUnplacedVolume const*, G4VSolid const*>
+        fUnplacedVolumeMap;
+    vecgeom::BidirectionalTypeMap<vecgeom::LogicalVolume const*,
+                                  G4LogicalVolume const*>
+        fLogicalVolumeMap;
+
+    // fast O(1) lookup to get VecGeom or G4 placed volume based on index
+    // FastG4VecGeomLookup fFastG4VGLookup;
+    std::vector<vecgeom::Transformation3D const*> fReplicaTransformations;
+
+    int fVerbose;
+
+  public:
+    /// Access singleton instance.
+    static G4VecGeomConverter& Instance()
+    {
+        static G4VecGeomConverter instance;
+        return instance;
+    }
```

Comment:
Singleton functions should definitely not be inline... and it's not clear to me why this should be a singleton at all rather than a locally constructed and destroyed helper class. Will it "own" any of the global vecgeom data it creates?